component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java

```diff
@@ -666,34 +666,18 @@ private boolean isValidTokenLocalUser(int token, String username, Authentication
             throws TOTPException {
 
         String tenantDomain = MultitenantUtils.getTenantDomain(username);
-        String tenantAwareUsername = null;
         try {
             TOTPAuthenticatorCredentials totpAuthenticator = getTotpAuthenticator(context, tenantDomain);
-            tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
-            UserRealm userRealm = TOTPUtil.getUserRealm(username);
-            if (userRealm != null) {
-                Map<String, String> userClaimValues = userRealm
-                        .getUserStoreManager().getUserClaimValues
-                                (tenantAwareUsername, new String[]
-                                        {TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL}, null);
-                String secretKey = TOTPUtil.decrypt(
-                        userClaimValues.get(TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL));
-                return totpAuthenticator.authorize(secretKey, token);
+            String secretKey;
+            if (context.getProperty(TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL) != null) {
+                secretKey = TOTPUtil.decrypt(
+                        context.getProperty(TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL).toString());
             } else {
-                throw new TOTPException(
-                        "Cannot find the user realm for the given tenant domain : " +
-                                CarbonContext.getThreadLocalCarbonContext().getTenantDomain());
+                throw new TOTPException("Error while retrieving the secret key");
             }
-        } catch (UserStoreException e) {
-            throw new TOTPException(
-                    "TOTPTokenVerifier failed while trying to access userRealm of the user : " +
-                            tenantAwareUsername, e);
+            return totpAuthenticator.authorizeAndStoreSecret(secretKey, token, username);
```

Comment:
I dont think we should update secret each time we call `isValidTokenLocalUser` method. It is violate the method name explanation. This method is used to validate the provided token of a local user against the secret key stored in the database (secretKey claim)

Reply:
This could cause TOTP validation failures as well because we are not setting the secret in the context in login flows

Reply:
Fixed